# PRAKTIKUM 1 #

In [1]:
import pandas as pd
import numpy as np

**Read Excel :** Untuk membaca data dari eksternal yaitu excel

In [ ]:
df = pd.read_csv("dataset_mahasiswa.csv")
df

,student_id,study_hours,attendance,assignments_completed,quiz_score,activity_level
0,1,12,95,10,92,Tinggi
1,2,10,90,9,88,Tinggi
2,3,8,85,8,82,Tinggi
3,4,6,75,7,74,Sedang
4,5,4,60,5,65,Sedang
5,6,2,45,3,52,Rendah
6,7,11,92,10,90,Tinggi
7,8,9,88,8,85,Tinggi
8,9,7,80,7,78,Sedang
9,10,5,70,6,68,Sedang


## Rumusan Masalah
| Jenis | Ide Analisis | Variabel |
|---|---|---|
| Deskriptif | Apakah terdapat perbedaan rata-rata nilai kuis antara mahasiswa dengan tingkat kehadiran rendah, sedang, dan tinggi? | `Attedance` |
| Eksploratif | Mahasiswa seperti apa yang perlu mendapatkan perhatian atau pendampingan berdasarkan jam belajar, kehadiran, penyelesaian tugas, dan nilai kuis? | `study_hour`, `attendance`, `assignments_completed`, `quiz_score`|
| Deskriptif | Berapa rata-rata jam belajar, kehadiran, jumlah tugas yang diselesaikan, dan nilai kuis mahasiswa? | `study_hours`, `attadance`, `assignments_completed`, `quiz_score` |
| Deskriptif | Bagaimana gambaran umum aktivitas belajar mahasiswa berdasarkan rata-rata jam belajar, persentase kehadiran, jumlah tugas yang diselesaikan, dan nilai kuis? | `study_hours`, `attendance`, `assignments_completed`, `quiz_score` |
| Prediktif | Sejauh mana jam belajar, persentase kehadiran, dan jumlah tugas yang diselesaikan dapat digunakan untuk memprediksi nilai kuis mahasiswa? | `study_hours`, `attedance`, `assignments_completed`, `quiz_score` |

In [6]:
# CODE UNTUK RUMUSAN MASALAH NO 3 

print("Rata-rata Jam Belajar:", df["study_hours"].mean())
print("Rata-rata Kehadiran:", df["attendance"].mean())
print("Rata-rata Tugas Diselesaikan:", df["assignments_completed"].mean())
print("Rata-rata Nilai Kuis:", df["quiz_score"].mean())

Rata-rata Jam Belajar: 6.633333333333334
Rata-rata Kehadiran: 73.56666666666666
Rata-rata Tugas Diselesaikan: 6.633333333333334
Rata-rata Nilai Kuis: 73.46666666666667


# Validasi #
1. **Apa yang ditemukan? :** 
Dapat dilihat bahwa rata rata Jam Belajar 6.63, Kehadiran 73.5, Tugas Diselesaikan 6.63, Nilai Kuis 73.4
2. **Apakah hasil sesuai dugaan awal? :** Sesuai seperti yang saya harapkan yaitu menunjukkan rata2 dan performa mahasiswa
3. **Keterbatasan analisis :** Data hanya 30 mahasiswa dan tidak menjelaskan kondisi setiap mahasiswa serta hubungan antar variabel
4. **Apakah diperlukan data tambahan :** Iya, seperti jenis matkul, nilai ujian, nilai tugas


In [ ]:
# Challenge #

def analyze_student(dataframe, student_id):
    try:
        # buat nyari nama Siswa pake student_id
        student = dataframe[dataframe["student_id"] == student_id]
    
        # cek apakah id siswa ada?
        if len(student) == 0:
            print("Siswa tidak ada")
        else: 
            study_hours = student["study_hours"].values[0]
            attendance = student["attendance"].values[0]
            quiz_score = student["quiz_score"].values[0]
        
        # menentukan status siswa
            if study_hours >= 6 and attendance >= 75 and quiz_score >= 70:
                status = "On Track"
                recommendation = "Tahanin belajar dan kehadiran."
            else:
                status = "Needs Review"
                recommendation = "Tingkatkan belajar atau kehadiran."
        
        #output 
        print("Student ID:", student_id)
        print("Jam Belajar:", study_hours, "jam")
        print("Kehadiran:", attendance, "%")
        print("Nilai Kuis:", quiz_score)
        print("Status:", status)
        print("Rekomendasi:", recommendation)
        
    except ValueError as e:
        print(e)
        
analyze_student(df, 1)

Student ID: 1
Jam Belajar: 12 jam
Kehadiran: 95 %
Nilai Kuis: 92
Status: On Track
Rekomendasi: Tahanin belajar dan kehadiran.


In [12]:
#Latihan

nama = "nicho"
umur = 20
tinggi = 160
mahasiswa = True
nilai = None

print(type(umur))
print(type(tinggi))
print(type(nama))
print(type(mahasiswa))
print(type(nilai))


<class 'int'>
<class 'int'>
<class 'str'>
<class 'bool'>
<class 'NoneType'>


# Refleksi #

Jawab dengan singkat:

1. **Apa perbedaan `list` Python dan NumPy array? :**
    list bisa menyimpan berbagai tipe data sedangkan numPy array untuk operasi numerik yg cepat

2. **Mengapa Pandas DataFrame banyak digunakan dalam data science? :** 
    memudahkan untuk pengelolahan dan analisis dalam bentuk tabel

3. **Kapan kita menggunakan dictionary? :** 
    saat menyimpan data dengan key-value terutama jika mengakses data berdasarkan nama/kunci

4. **Apa manfaat function dibanding menulis kode berulang? :** 
    kode menjadi lebih terstruktur dan mengurangi pengulangan

5. **Apa fungsi `try-except`? :** 
    agar program tidak langsung berhenti ketika ada salah

6. **Apa perbedaan iterator dan generator? :** 
    iterator buat mengambil data 1 per 1 kalo generator itu cara ngebuat iterator 

7. **Mengapa validasi data penting dalam pipeline? :** 
    Memastikan data benar dan sesuai format agar hasil analisis bisa dipercaya

8. **Dalam studi kasus ini, pada bagian mana AI membantu menghasilkan atau menstrukturkan ide? :** 
    bagian ide pertanyaan / rumusan masalah dan menghasilkan datanya

9. **Mengapa hasil yang diberikan AI tetap harus divalidasi menggunakan data dan kode? :** 
    karna ai dapat menghasilkan kode maupun informasi yg kurang tepat

10. **Bagaimana konsep SDLC terlihat dalam mini pipeline praktikum ini? :** 
    masalah -> Data -> Ekplorasi -> analisis -> Insight -> rekomendasi -> validasi 

## 22. Checklist Kompetensi

Centang setelah mampu melakukan secara mandiri:

- [✓] Import NumPy dan Pandas
- [✓] Membuat variabel
- [✓] Mengenali tipe data Python
- [✓] Melakukan operasi dasar
- [✓] Menggunakan list
- [ ] Menggunakan NumPy vector/array
- [✓] Menggunakan dictionary
- [✓] Membuat DataFrame
- [✓] Membaca CSV
- [✓] Melakukan filtering Pandas
- [✓] Membuat function
- [✓] Menggunakan conditional
- [✓] Menangani exception
- [✓] Membuat class dan object
- [ ] Menggunakan iterator
- [ ] Membuat generator
- [ ] Menggabungkan konsep dalam pipeline sederhana
- [✓] Menggunakan AI untuk menghasilkan dan menstrukturkan ide analisis
- [✓] Memvalidasi hasil AI menggunakan data dan kode